incrementalprune.go

```diff
@@ -0,0 +1,764 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package txn
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/lru"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const pruneTxnBatchSize = 1000
+const pruneMinTxnBatchSize = 10
+const pruneMaxTxnBatchSize = 10000
+const defaultBatchSleepTime time.Duration = 0
+const maxBatchSleepTime = 1 * time.Second
+const queryDocBatchSize = 100
+const pruneDocCacheSize = 10000
+const missingKeyCacheSize = 2000
+const strCacheSize = 10000
+
+// IncrementalPruner reads the transaction table incrementally, seeing if it can remove the current set of transactions,
+// and then moves on to newer transactions. It only thinks about 1k txns at a time, because that is the batch size that
+// can be deleted. Instead, it caches documents that it has seen.
```

Comment:
The original prune code would grab the list of all known transaction ids, then walk all of the documents in the database, removing from the 'safe' set the list of ids that were still referenced. After walking everything it would then walk the list of remaining transaction ids as safe to remove.

Which is a 'safe' strategy, but requires doing a lot of work before you can do anything.